src/vs/workbench/contrib/terminal/common/scripts/shellIntegration-bash.sh

```diff
@@ -214,16 +217,77 @@ __vsc_update_cwd() {
 	builtin printf '\e]633;P;Cwd=%s\a' "$(__vsc_escape_value "$__vsc_cwd")"
 }
 
-# __vsc_update_env() {
-# 	builtin printf '\e]633;EnvSingleStart;%s;\a' $__vsc_nonce
-# 	for var in $(compgen -v); do
-# 		if printenv "$var" >/dev/null 2>&1; then
-# 			value=$(builtin printf '%s' "${!var}")
-# 			builtin printf '\e]633;EnvSingleEntry;%s;%s;%s\a' "$var" "$(__vsc_escape_value "$value")" $__vsc_nonce
-# 		fi
-# 	done
-# 	builtin printf '\e]633;EnvSingleEnd;%s;\a' $__vsc_nonce
-# }
+
+updateEnvCache() {
+	local key="$1"
+	local value="$2"
+
+	for i in "${!vsc_env_keys[@]}"; do
+		if [[ "${vsc_env_keys[$i]}" == "$key" ]]; then
+			if [[ "${vsc_env_values[$i]}" != "$value" ]]; then
+				vsc_env_values[$i]="$value"
+				# printf "Updated value for %s\n" "$key"
+				builtin printf '\e]633;EnvSingleEntry;%s;%s;%s\a' "$key" "$(__vsc_escape_value "$value")" "$__vsc_nonce"
+			fi
+			return
+		fi
+	done
+
+	vsc_env_keys+=("$key")
+	vsc_env_values+=("$value")
+	# printf "Added new variable %s\n" "$key"
+	builtin printf '\e]633;EnvSingleEntry;%s;%s;%s\a' "$key" "$(__vsc_escape_value "$value")" "$__vsc_nonce"
+}
+
+trackMissingEnvVars() {
+	# Capture current environment variables in an array
+	local current_env_keys=()
+	while IFS='=' read -r key value; do
+		current_env_keys+=("$key")
+	done < <(env)
+
+	# Compare vsc_env_keys with current_env_keys
+	for key in "${vsc_env_keys[@]}"; do
+		local found=false
+		for env_key in "${current_env_keys[@]}"; do
+			if [[ "$key" == "$env_key" ]]; then
+				found=true
+				break
+			fi
+		done
+		if [[ "$found" == false ]]; then
+			printf "Missing environment variable: %s\n" "$key"
+			unset 'vsc_env_keys[i]'
+			unset 'vsc_env_values[i]'
+		fi
+	done
+
+	# Remove gaps from unset
+	vsc_env_keys=("${vsc_env_keys[@]}")
+	vsc_env_values=("${vsc_env_values[@]}")
+}
+
+__vsc_update_env() {
+
+	start_time=$(date +%s)
+	builtin printf "Start time: %s\n" "$start_time"
+
+	builtin printf '\e]633;EnvSingleStart;%s;\a' $__vsc_nonce
+
+	while IFS='=' read -r key value; do
+		updateEnvCache "$key" "$value"
+	done < <(env)
+
+	builtin printf '\e]633;EnvSingleEnd;%s;\a' $__vsc_nonce
+
+	end_time=$(date +%s)
+	elapsed_time=$(($end_time - $start_time))
+
+	# builtin printf "Time taken: $elapsed_time seconds\n"
+
+	trackMissingEnvVars
+	builtin printf "Time taken: $elapsed_time seconds\n"
```

Comment:
Although we have more loops now in the script, time is actually shorter because we use the least number of printf statement as much as possible, avoid using printenv, and compgen. 

Check this by looking at the time calculation and just how things feel.